src/nvim/autocmd.c

```diff
@@ -2117,3 +2117,29 @@ bool au_exists(const char *const arg) FUNC_ATTR_WARN_UNUSED_RESULT
   xfree(arg_save);
   return retval;
 }
+
+void init_default_autocmds(void)
+{
+  // open terminals when opening files that start with term://
+#define PROTO "term://"
+  do_augroup((char_u *)"nvim_terminal", false);
+  do_autocmd_event(EVENT_BUFREADCMD, (char_u *)PROTO "*", false, true,
+                   (char_u *)":if !exists('b:term_title')|call termopen( "
+                   // Capture the command string
+                   "matchstr(expand(\"<amatch>\"), "
+                   "'\\c\\m" PROTO "\\%(.\\{-}//\\%(\\d\\+:\\)\\?\\)\\?\\zs.*'), "
+                   // capture the working directory
+                   "{'cwd': expand(get(matchlist(expand(\"<amatch>\"), "
+                   "'\\c\\m" PROTO "\\(.\\{-}\\)//'), 1, ''))})"
+                   "|endif",
+                   true,
+                   AUGROUP_ALL);
+  do_augroup((char_u *)"END", false);
+#undef PROTO
+
+  // Only sync 1 syntax line in the command window
+  do_augroup((char_u *)"nvim_cmdwin", false);
```

Comment:
need to document this augroup somewhere . need to think about this.

also wonder if we can avoid all the C string junk and put this in a "special" runtime file that is sourced by main() or whatever. That also helps discoverability and easier inspection, we can point to that runtime file from the docs.

Reply:
> also wonder if we can avoid all the C string junk and put this in a "special" runtime file that is sourced by main() or whatever. That also helps discoverability and easier inspection, we can point to that runtime file from the docs.

That's a good idea. I think in this case the discoverability and maintainability of having it in its own runtime file outweigh whatever performance benefits come from doing it directly in C.

Reply:
> We could put all default autocmds into a single group and then just document that?

We want users to be able to selectively opt-out of various defaults. A single augroup defeats that. I think `nvim_*` augroup name convention is OK for now, haven't thought of something better. (Lua will make this better: we can group events in a global map.)

Reply:
One "gotcha" with using a runtime file is that it changes the script IDs of all subsequent scripts. This causes a lot of the tests to fail (since they assume script IDs start at 1, but they now start at 2) in addition to making the user's init.vim no longer have SID 1.

Possible solutions:

1. Don't execute the autocmd script when using `-u NONE`. This would fix the tests at least, but would prevent the `term://` BufReadCmd autocmd from working with `-u NONE`
2. Source the autocmds.vim file after the user's init.vim. This would make init.vim still have SID 1, but the user can no longer remove these autocmds from their init.vim if they want to (I suppose we could add a guard around autocmds.vim so that user's can prevent it from loading). This would also prevent the `term://` BufReadCmd from working properly with e.g. `-c 'e term://ls'` on the command line.

My 2 cents is that we do both (1) and (2) but keep the `nvim_terminal` autocommand as it is now, defined directly in `main.c` and only use `autocmds.vim` for the CmdWinEnter autocommand and any future default autocommands. Thoughts?

Reply:
> This causes a lot of the tests to fail (since they assume script IDs start at 1, but they now start at 2)

I'm surprised by this, and that's arguably a bug in the tests.  Upstream's test infrastructure sources numerous scripts as part of running the tests, so this sounds like something specific to _our_ tests.

> in addition to making the user's init.vim no longer have SID 1.

Meh, this just sounds like a case of [spacebar heating](https://xkcd.com/1172).

> My 2 cents is that we do both (1) and (2) but keep the `nvim_terminal` autocommand as it is now, defined directly in `main.c` and only use `autocmds.vim` for the CmdWinEnter autocommand and any future default autocommands. Thoughts?

Sounds good to me.  The important part is to be consistent.  I was starting to get flashbacks to the whole defaults.vim debacle, but thankfully you decided on always loading the file while providing the user the capability of short-circuiting the contents.